internal/node/node.go

```diff
@@ -0,0 +1,300 @@
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/nats-io/nats-server/v2/server"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nkeys"
+	"github.com/synadia-io/nex/internal/models"
+)
+
+const defaultNatsStoreDir = "pnats"
+
+const runloopSleepInterval = 250 * time.Millisecond
+const runloopTickInterval = 2500 * time.Millisecond
+
+// Nex node process
+type Node struct {
+	api     *ApiListener
+	manager *MachineManager
+
+	cancelF context.CancelFunc
+	closing uint32
+	ctx     context.Context
+	sigs    chan os.Signal
+
+	log *slog.Logger
+
+	config   *NodeConfiguration
+	opts     *models.Options
+	nodeOpts *models.NodeOptions
+
+	initOnce sync.Once
+
+	keypair   nkeys.KeyPair
+	publicKey string
+
+	nc *nats.Conn
+
+	natsint *server.Server
+	ncint   *nats.Conn
+
+	telemetry *Telemetry
+}
+
+func NewNode(opts *models.Options, nodeOpts *models.NodeOptions, ctx context.Context, cancelF context.CancelFunc, log *slog.Logger) (*Node, error) {
+	node := &Node{
+		ctx:      ctx,
+		cancelF:  cancelF,
+		log:      log,
+		nodeOpts: nodeOpts,
+		opts:     opts,
+	}
+
+	err := node.validateConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start node: %s", err.Error())
+	}
+
+	err = node.generateKeypair()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate keypair for node: %s", err)
+	} else {
+		node.log.Info("Generated keypair for node", slog.String("public_key", node.publicKey))
+	}
+
+	return node, nil
+}
+
+func (n *Node) PublicKey() (*string, error) {
+	pubkey, err := n.keypair.PublicKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pubkey, nil
+}
+
+func (n *Node) Start() {
+	n.log.Debug("starting node", slog.String("public_key", n.publicKey))
+
+	err := n.init()
+	if err != nil {
+		panic(err) // FIXME-- this panics here because this is written like a proper main() entrypoint (it should never actually panic in practice)
+	}
+
+	// startAt := time.Now()
+
+	timer := time.NewTicker(runloopTickInterval)
+	defer timer.Stop()
+
+	for !n.shuttingDown() {
+		select {
+		case <-timer.C:
+			// TODO: check NATS subscription statuses, machine manager, telemetry etc.
+		case sig := <-n.sigs:
+			n.log.Debug("received signal: %s", sig)
+			n.shutdown()
+		case <-n.ctx.Done():
+			close(n.sigs)
+		default:
+			time.Sleep(runloopSleepInterval)
+		}
+	}
+
+	n.log.Info("exiting node")
+	n.cancelF()
+}
+
+func (n *Node) Stop() {
+	n.log.Debug("stopping node")
+	n.shutdown()
+}
+
+func (n *Node) generateKeypair() error {
+	var err error
+
+	n.keypair, err = nkeys.CreateServer()
+	if err != nil {
+		return fmt.Errorf("failed to generate node keypair: %s", err)
+	}
+
+	n.publicKey, err = n.keypair.PublicKey()
+	if err != nil {
+		return fmt.Errorf("failed to encode public key: %s", err)
+	}
+
+	return nil
+}
+
+func (n *Node) init() error {
+	var err error
+
+	n.initOnce.Do(func() {
+		if n.nodeOpts.OtelMetrics {
+			n.telemetry, err = NewTelemetry(n.log, n.nodeOpts.OtelMetricsExporter, n.nodeOpts.OtelMetricsPort)
```

Comment:
We should create the Provider here and then do NewTelemety.  That way, Telemetry is never nil and the opt-in is whether or not the telemetry is propagated

Reply:
I kept the logic in telemetry.go but otherwise believe to have captured your feedback.